server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java

```diff
@@ -275,14 +277,72 @@ public static void runJobInTransaction(KeycloakSessionFactory factory, KeycloakC
     }
 
     /**
-     * Copy all the objects in the context to the session.
-     * @param session The session
-     * @param context The context
+     * Sets up the context for the specified session with the RealmModel.
+     *
+     * @param origContext The original context to propagate
+     * @param targetSession The new target session to propagate the context to
+     */
+    public static void cloneContextRealmClientToSession(final KeycloakContext origContext, final KeycloakSession targetSession) {
+        cloneContextToSession(origContext, targetSession, false);
+    }
+
+    /**
+     * Sets up the context for the specified session with the RealmModel, clientModel and
+     * AuthenticatedSessionModel.
+     *
+     * @param origContext The original context to propagate
+     * @param targetSession The new target session to propagate the context to
      */
-    public static void propagateContext(KeycloakSession session, KeycloakContext context) {
-        session.getContext().setRealm(context.getRealm());
-        session.getContext().setClient(context.getClient());
-        session.getContext().setAuthenticationSession(context.getAuthenticationSession());
+    public static void cloneContextRealmClientSessionToSession(final KeycloakContext origContext, final KeycloakSession targetSession) {
+        cloneContextToSession(origContext, targetSession, true);
+    }
+
+    /**
+     * Sets up the context for the specified session.The original realm's context is used to
+     * determine what models need to be re-loaded using the current session. The models
+     * in the context are re-read from the new session via the IDs.
+     */
+    private static void cloneContextToSession(final KeycloakContext origContext, final KeycloakSession targetSession,
+            final boolean includeAuthenticatedSessionModel) {
+        if (origContext == null) {
+            return;
+        }
+
+        // setup realm model if necessary.
+        if (origContext.getRealm() == null) {
+            return;
```

Comment:
Please address https://github.com/rmartinc/keycloak/commit/a7eafec4da46d97a6fa88c7eae31e6113508a941#r117140224

Reply:
Done! Now the three model objects are read dependently.